File: scripts/analysis/recalculate_v7_key_results.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2L) stop("Usage: recalculate_v7_key_results.R <root> <outdir>")

root <- normalizePath(args[[1]], winslash = "/", mustWork = TRUE)
outdir <- args[[2]]
dir.create(outdir, recursive = TRUE, showWarnings = FALSE)

suppressPackageStartupMessages(library(survival))
suppressPackageStartupMessages(library(jsonlite))

read_tsv <- function(path) read.delim(path, check.names = FALSE, stringsAsFactors = FALSE)
as_num <- function(x) as.numeric(x)

cox_summary <- function(fit, module_name = "module_z") {
  s <- summary(fit)
  coef_row <- s$coefficients[module_name, ]
  ci_row <- s$conf.int[module_name, ]
  zph <- cox.zph(fit, transform = "rank")$table
  list(
    n = unname(fit$n),
    events = unname(fit$nevent),
    hr = unname(ci_row[["exp(coef)"]]),
    ci95_low = unname(ci_row[["lower .95"]]),
    ci95_high = unname(ci_row[["upper .95"]]),
    p = unname(coef_row[["Pr(>|z|)"]]),
    module_ph_p_rank = unname(zph[module_name, "p"]),
    global_ph_p_rank = unname(zph["GLOBAL", "p"])
  )
}

results <- list()

# Organoid anchors and subgroup tests
org <- file.path(root, "analysis", "organoid")
panel <- read.csv(file.path(org, "organoid_panel_dependency_summary.csv"), check.names = FALSE)
screen <- read.csv(gzfile(file.path(org, "organoid_crc_best_library_screen.csv.gz")), check.names = FALSE)
ann <- read.csv(file.path(org, "organoid_model_annotations.csv"), check.names = FALSE)
screen <- merge(screen, ann[, c("sample_ID", "individual_ID", "msStatus", "CMS_prediction", "KRAS_class")],
                by = "sample_ID", all.x = TRUE, sort = FALSE)

results$organoid_anchors <- lapply(c("SCD", "EGFR"), function(gene) {
  values <- screen$LFC[screen$gene == gene]
  stored <- panel[panel$gene == gene, ]
  list(gene = gene, n = sum(!is.na(values)), median_lfc_recomputed = median(values, na.rm = TRUE),
       median_lfc_stored = stored$median_LFC[[1]], official_pct_depleted = stored$pct_depleted[[1]])
})

cms_order <- c("CMS1", "CMS2", "CMS3", "CMS4")
scd <- screen[screen$gene == "SCD", ]
cms_groups <- lapply(cms_order, function(cms) scd$LFC[scd$CMS_prediction == cms & !is.na(scd$CMS_prediction)])
results$organoid_scd_cms <- list(
  counts = sapply(cms_groups, length),
  medians = sapply(cms_groups, median, na.rm = TRUE),
  kruskal_p = kruskal.test(cms_groups)$p.value
)

msi_rows <- lapply(sort(unique(panel$gene)), function(gene) {
  subset <- screen[screen$gene == gene, ]
  msi <- subset$LFC[subset$msStatus == "MSI"]
  mss <- subset$LFC[subset$msStatus == "MSS"]
  data.frame(gene = gene, n_msi = sum(!is.na(msi)), n_mss = sum(!is.na(mss)),
             p = wilcox.test(msi, mss, exact = FALSE, correct = FALSE)$p.value)
})
msi_table <- do.call(rbind, msi_rows)
msi_table$q <- p.adjust(msi_table$p, method = "BH")
results$organoid_msi <- list(
  n_msi = sum(ann$msStatus == "MSI", na.rm = TRUE),
  n_mss = sum(ann$msStatus == "MSS", na.rm = TRUE),
  genes = lapply(c("SCD", "EGFR"), function(gene) as.list(msi_table[msi_table$gene == gene, ])),
  n_fdr_lt_005 = sum(msi_table$q < 0.05)
)

# DepMap cross-platform calculations
dep <- file.path(root, "analysis", "depmap_26Q1")
cross <- read.csv(file.path(dep, "depmap_organoid_cross_platform_panel.csv"), check.names = FALSE)
full_cor <- cor.test(cross$median_organoid, cross$median_crc_2d, method = "spearman", exact = FALSE)
legacy <- cross[cross$gene != "CD274", ]
legacy_cor <- cor.test(legacy$median_organoid, legacy$median_crc_2d, method = "spearman", exact = FALSE)
context <- read.csv(file.path(dep, "depmap_target_context_summary.csv"), check.names = FALSE)
anchor_context <- lapply(c("SCD", "EGFR"), function(gene) {
  row <- context[context$gene == gene, ]
  as.list(row[, c("gene", "median_crc_2d", "pct_crc_lt_minus_1", "median_difference_crc_minus_other",
                  "median_difference_ci95_lo", "median_difference_ci95_hi", "mannwhitney_p_crc_vs_other")])
})
results$cross_platform <- list(
  n_full = nrow(cross), rho_full = unname(full_cor$estimate), p_full = full_cor$p.value,
  n_legacy = nrow(legacy), rho_legacy = unname(legacy_cor$estimate), p_legacy = legacy_cor$p.value,
  anchors = anchor_context
)

# SCD co-dependency and Hallmark sensitivity outputs
codep <- file.path(root, "analysis", "scd_codependency_qc")
rank <- read.csv(file.path(codep, "scd_codependency_qc_adjusted.csv"), check.names = FALSE)
gsea <- read.csv(file.path(codep, "scd_hallmark_raw_vs_qc.csv"), check.names = FALSE)
target_rows <- rank[rank$gene %in% c("VPS72", "CDX1", "WLS"), ]
results$scd_codependency <- list(
  n_genes = nrow(rank),
  n_gene_fdr_lt_005 = sum(rank$FDR_BH < 0.05),
  target_rows = split(target_rows, target_rows$gene),
  n_hallmark_fdr_lt_005 = sum(gsea$FDR_BH < 0.05),
  dna_repair = split(gsea[gsea$pathway == "HALLMARK_DNA_REPAIR", ], gsea$ranking[gsea$pathway == "HALLMARK_DNA_REPAIR"]),
  cholesterol_homeostasis = split(gsea[gsea$pathway == "HALLMARK_CHOLESTEROL_HOMEOSTASIS", ], gsea$ranking[gsea$pathway == "HALLMARK_CHOLESTEROL_HOMEOSTASIS"])
)

# Survival recalculation from frozen patient-level tables
gse <- read_tsv(file.path(root, "analysis", "gse39582_recalculation", "gse39582_module_clinical.tsv"))
gse_univ_data <- gse[gse$rfs_months > 0 & complete.cases(gse[, c("rfs_months", "rfs_event", "module_z")]), ]
gse_adj_vars <- c("rfs_months", "rfs_event", "module_z", "age", "mmrd", "chemotherapy_yes", "stage_2", "stage_3", "stage_4")
gse_adj_data <- gse[gse$rfs_months > 0 & complete.cases(gse[, gse_adj_vars]), ]
gse_univ_fit <- coxph(Surv(rfs_months, rfs_event) ~ module_z, data = gse_univ_data, ties = "efron", x = TRUE)
gse_adj_fit <- coxph(Surv(rfs_months, rfs_event) ~ module_z + age + mmrd + chemotherapy_yes + stage_2 + stage_3 + stage_4,
                     data = gse_adj_data, ties = "efron", x = TRUE)
results$gse39582 <- list(rfs_univariable = cox_summary(gse_univ_fit), rfs_adjusted = cox_summary(gse_adj_fit), os_analysable = FALSE)

tcga <- read_tsv(file.path(root, "analysis", "tcga_cbioportal_592", "tcga_cbioportal_module_clinical.tsv"))
tcga$stage_factor <- factor(tcga$stage_num, levels = c(1, 2, 3, 4))
tcga_univ_data <- tcga[tcga$os_months > 0 & complete.cases(tcga[, c("os_months", "os_event", "module_z")]), ]
tcga_adj_data <- tcga[tcga$os_months > 0 & complete.cases(tcga[, c("os_months", "os_event", "module_z", "age", "stage_factor")]), ]
tcga_univ_fit <- coxph(Surv(os_months, os_event) ~ module_z, data = tcga_univ_data, ties = "efron", x = TRUE)
tcga_adj_fit <- coxph(Surv(os_months, os_event) ~ module_z + age + stage_factor, data = tcga_adj_data, ties = "efron", x = TRUE)
results$tcga <- list(expression_n = sum(!is.na(tcga$module_z)), os_univariable = cox_summary(tcga_univ_fit),
                     os_age_stage_adjusted = cox_summary(tcga_adj_fit))

# Patient-paired candidate-cell recalculation
leuko <- file.path(root, "analysis", "gse146771")
rates <- read.csv(file.path(leuko, "gse146771_candidate_rates_by_patient_tissue_lineage.csv"), check.names = FALSE)
all_rates <- rates[rates$lineage == "All leukocytes", ]
wide <- reshape(all_rates[, c("Sample", "Tissue", "rate")], idvar = "Sample", timevar = "Tissue", direction = "wide")
paired <- wide[complete.cases(wide[, c("rate.N", "rate.T")]), ]
differences <- paired$rate.T - paired$rate.N
sign_matrix <- expand.grid(rep(list(c(-1, 1)), length(differences)))
permuted <- apply(sign_matrix, 1, function(s) abs(mean(differences * as.numeric(s))))
sign_p <- mean(permuted >= abs(mean(differences)) - 1e-15)
tissue <- read.csv(file.path(leuko, "gse146771_candidate_tissue_descriptive.csv"), check.names = FALSE)
sensitivity <- read.csv(file.path(leuko, "gse146771_threshold_sensitivity.csv"), check.names = FALSE)
results$gse146771 <- list(
  paired_patients = nrow(paired), mean_difference_T_minus_N = mean(differences), exact_sign_flip_p = sign_p,
  tissue_counts = split(tissue, tissue$tissue),
  threshold_candidate_range = lapply(split(sensitivity$total_candidates, sensitivity$CEACAM5_TPM_threshold), range)
)

# Patient-level GSE178341 recalculation
g178 <- read_tsv(file.path(root, "analysis", "gse178341_recalculation", "gse178341_patient_scores.tsv"))
cms4 <- g178$prediction_raw == "CMS4"
cor_pairs <- list(
  module_all_vs_TGFb_all = c("module_all", "TGFb_all"),
  module_all_vs_EMT_all = c("module_all", "EMT_all"),
  module_Epi_vs_TGFb_Epi = c("module_Epi", "TGFb_Epi"),
  module_Epi_vs_EMT_Epi = c("module_Epi", "EMT_Epi"),
  module_all_vs_T_cell_fraction_all = c("module_all", "T_cell_fraction_all"),
  module_Epi_vs_T_cell_fraction_all = c("module_Epi", "T_cell_fraction_all")
)
correlations <- lapply(cor_pairs, function(pair) {
  test <- cor.test(g178[[pair[[1]]]], g178[[pair[[2]]]], method = "spearman", exact = FALSE)
  list(rho = unname(test$estimate), p = test$p.value)
})
mmrd <- g178$MMRStatus == "MMRd"
mmr_columns <- c("T_cell_fraction_all", "exhausted_label_fraction_T", "exhaustion_expression_TNKILC")
mmr_tests <- lapply(mmr_columns, function(column) {
  test <- wilcox.test(g178[[column]][mmrd], g178[[column]][!mmrd], exact = FALSE, correct = FALSE)
  list(variable = column, median_mmrd = median(g178[[column]][mmrd], na.rm = TRUE),
       median_mmrp = median(g178[[column]][!mmrd], na.rm = TRUE), p = test$p.value)
})
fdr_classified <- !is.na(g178$prediction_FDR05) & g178$prediction_FDR05 != "Unclassified"
cms_p_values <- c(
  wilcox.test(g178$module_all[cms4], g178$module_all[!cms4], exact = FALSE, correct = FALSE)$p.value,
  wilcox.test(g178$module_all[g178$prediction_FDR05 == "CMS4"],
              g178$module_all[fdr_classified & g178$prediction_FDR05 != "CMS4"], exact = FALSE, correct = FALSE)$p.value,
  wilcox.test(g178$module_Epi[cms4], g178$module_Epi[!cms4], exact = FALSE, correct = FALSE)$p.value,
  wilcox.test(g178$module_Epi[g178$prediction_FDR05 == "CMS4"],
              g178$module_Epi[fdr_classified & g178$prediction_FDR05 != "CMS4"], exact = FALSE, correct = FALSE)$p.value
)
group_q <- p.adjust(c(sapply(mmr_tests, function(x) x$p), cms_p_values), method = "BH")
for (i in seq_along(mmr_tests)) mmr_tests[[i]]$q <- group_q[[i]]
results$gse178341 <- list(
  patients = length(unique(g178$PID)), mmrd = sum(mmrd), mmrp = sum(!mmrd),
  cms_raw_counts = as.list(table(g178$prediction_raw)),
  cms4_whole_p = cms_p_values[[1]], cms4_whole_q = group_q[[4]],
  cms4_whole_fdr_call_p = cms_p_values[[2]], cms4_whole_fdr_call_q = group_q[[5]],
  cms4_epithelial_p = cms_p_values[[3]], cms4_epithelial_q = group_q[[6]],
  cms4_epithelial_fdr_call_p = cms_p_values[[4]], cms4_epithelial_fdr_call_q = group_q[[7]],
  correlations = correlations, mmr_tests = mmr_tests
)

output <- file.path(outdir, "v7_key_results_recalculated.json")
write_json(results, output, auto_unbox = TRUE, pretty = TRUE, digits = 16, na = "null")
cat(output, "\n")
cat(toJSON(results, auto_unbox = TRUE, pretty = TRUE, digits = 8, na = "null"), "\n")

File: scripts/analysis/recalculate_v7_key_results.py
from __future__ import annotations

import argparse
import itertools
import json
from pathlib import Path

import numpy as np
import pandas as pd
from scipy import optimize, stats


def bh_adjust(p_values: list[float] | np.ndarray) -> np.ndarray:
    p = np.asarray(p_values, dtype=float)
    order = np.argsort(p)
    ranked = p[order]
    adjusted = ranked * len(p) / np.arange(1, len(p) + 1)
    adjusted = np.minimum.accumulate(adjusted[::-1])[::-1]
    output = np.empty_like(adjusted)
    output[order] = np.clip(adjusted, 0.0, 1.0)
    return output


def cox_components(beta: np.ndarray, x: np.ndarray, time: np.ndarray, event: np.ndarray) -> tuple[float, np.ndarray, np.ndarray]:
    """Efron partial log-likelihood, gradient and Hessian."""
    eta = np.clip(x @ beta, -50.0, 50.0)
    weight = np.exp(eta)
    p = x.shape[1]
    loglik = 0.0
    gradient = np.zeros(p)
    hessian = np.zeros((p, p))
    for event_time in np.unique(time[event == 1]):
        deaths = (time == event_time) & (event == 1)
        risk = time >= event_time
        d = int(deaths.sum())
        x_death = x[deaths]
        w_risk = weight[risk]
        x_risk = x[risk]
        w_death = weight[deaths]
        s0 = float(w_risk.sum())
        s1 = np.einsum("i,ij->j", w_risk, x_risk)
        s2 = np.einsum("i,ij,ik->jk", w_risk, x_risk, x_risk)
        e0 = float(w_death.sum())
        e1 = np.einsum("i,ij->j", w_death, x_death)
        e2 = np.einsum("i,ij,ik->jk", w_death, x_death, x_death)
        # Sum tied-event linear predictors before converting to a scalar.
        # The former ``float(array).sum()`` form fails under current NumPy.
        loglik += float((x_death @ beta).sum())
        gradient += x_death.sum(axis=0)
        for fraction_index in range(d):
            fraction = fraction_index / d
            denom = s0 - fraction * e0
            first = s1 - fraction * e1
            second = s2 - fraction * e2
            loglik -= np.log(denom)
            gradient -= first / denom
            hessian -= second / denom - np.outer(first, first) / (denom * denom)
    return loglik, gradient, hessian


def approximate_schoenfeld_p(beta: np.ndarray, x: np.ndarray, time: np.ndarray, event: np.ndarray) -> np.ndarray:
    eta = np.clip(x @ beta, -50.0, 50.0)
    weight = np.exp(eta)
    residuals: list[np.ndarray] = []
    event_times: list[float] = []
    for event_time in np.unique(time[event == 1]):
        deaths = (time == event_time) & (event == 1)
        risk = time >= event_time
        d = int(deaths.sum())
        x_death = x[deaths]
        w_risk, x_risk = weight[risk], x[risk]
        w_death = weight[deaths]
        s0 = float(w_risk.sum())
        s1 = np.einsum("i,ij->j", w_risk, x_risk)
        e0 = float(w_death.sum())
        e1 = np.einsum("i,ij->j", w_death, x_death)
        expected = np.mean([(s1 - (j / d) * e1) / (s0 - (j / d) * e0) for j in range(d)], axis=0)
        for row in x_death:
            residuals.append(row - expected)
            event_times.append(float(event_time))
    resid = np.vstack(residuals)
    ranked_time = stats.rankdata(np.asarray(event_times))
    p_values = []
    for column in range(resid.shape[1]):
        test = stats.pearsonr(ranked_time, resid[:, column])
        p_values.append(float(test.pvalue))
    return np.asarray(p_values)


def cox_fit(frame: pd.DataFrame, duration: str, event: str, covariates: list[str]) -> dict[str, float | int]:
    data = frame[[duration, event, *covariates]].dropna().copy()
    data = data[data[duration] > 0].copy()
    x = data[covariates].astype(float).to_numpy()
    time = data[duration].astype(float).to_numpy()
    event_values = data[event].astype(int).to_numpy()
    scale = x.std(axis=0, ddof=0)
    scale[scale == 0] = 1.0
    centre = x.mean(axis=0)
    x_scaled = (x - centre) / scale

    def objective(beta: np.ndarray) -> tuple[float, np.ndarray, np.ndarray]:
        ll, grad, hess = cox_components(beta, x_scaled, time, event_values)
        return -ll, -grad, -hess

    fitted = optimize.minimize(
        lambda beta: objective(beta)[0], np.zeros(x_scaled.shape[1]),
        jac=lambda beta: objective(beta)[1], hess=lambda beta: objective(beta)[2],
        method="trust-exact", options={"gtol": 1e-9, "maxiter": 500},
    )
    if not fitted.success:
        # SciPy 1.18 may reject an otherwise well-behaved trust-region step
        # when the Hessian approximation cannot predict improvement. BFGS
        # uses the same Efron log-likelihood and analytic gradient and is a
        # stable compatibility fallback.
        fitted = optimize.minimize(
            lambda beta: objective(beta)[0], np.zeros(x_scaled.shape[1]),
            jac=lambda beta: objective(beta)[1], method="BFGS",
            options={"gtol": 1e-8, "maxiter": 2000},
        )
    gradient_norm = float(np.linalg.norm(objective(fitted.x)[1], ord=np.inf))
    if not fitted.success and gradient_norm > 1e-6:
        raise RuntimeError(
            f"Cox fit failed: {fitted.message}; gradient_inf={gradient_norm:.3g}"
        )
    beta_scaled = fitted.x
    _, _, hess_scaled = cox_components(beta_scaled, x_scaled, time, event_values)
    covariance_scaled = np.linalg.pinv(-hess_scaled)
    beta = beta_scaled / scale
    covariance = covariance_scaled / np.outer(scale, scale)
    se = np.sqrt(np.diag(covariance))
    module_index = covariates.index("module_z")
    z = beta[module_index] / se[module_index]
    p_value = float(2 * stats.norm.sf(abs(z)))
    ph = approximate_schoenfeld_p(beta_scaled, x_scaled, time, event_values)
    return {
        "n": int(len(data)),
        "events": int(event_values.sum()),
        "hr": float(np.exp(beta[module_index])),
        "ci95_low": float(np.exp(beta[module_index] - 1.96 * se[module_index])),
        "ci95_high": float(np.exp(beta[module_index] + 1.96 * se[module_index])),
        "p": p_value,
        "module_ph_p_rank_approx": float(ph[module_index]),
        "global_ph_min_p_rank_approx": float(ph.min()),
    }


def exact_sign_flip_p(differences: np.ndarray) -> float:
    differences = np.asarray(differences, dtype=float)
    observed = abs(float(differences.mean()))
    values = []
    for signs in itertools.product((-1.0, 1.0), repeat=len(differences)):
        values.append(abs(float(np.mean(differences * np.asarray(signs)))))
    return float(np.mean(np.asarray(values) >= observed - 1e-15))


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--root", required=True, type=Path)
    parser.add_argument("--outdir", required=True, type=Path)
    args = parser.parse_args()
    root = args.root
    args.outdir.mkdir(parents=True, exist_ok=True)
    results: dict[str, object] = {}

    org = root / "analysis" / "organoid"
    panel = pd.read_csv(org / "organoid_panel_dependency_summary.csv")
    screen = pd.read_csv(org / "organoid_crc_best_library_screen.csv.gz")
    ann = pd.read_csv(org / "organoid_model_annotations.csv")
    screen = screen.merge(
        ann[["sample_ID", "individual_ID", "msStatus", "CMS_prediction", "KRAS_class"]],
        on="sample_ID", how="left", validate="many_to_one"
    )
    anchor_rows: list[dict[str, object]] = []
    for gene in ("SCD", "EGFR"):
        values = screen.loc[screen["gene"] == gene, "LFC"].dropna()
        row = panel.set_index("gene").loc[gene]
        anchor_rows.append({
            "gene": gene,
            "n": int(len(values)),
            "median_lfc_recomputed": float(values.median()),
            "median_lfc_stored": float(row["median_LFC"]),
            "official_pct_depleted": float(row["pct_depleted"]),
        })
    results["organoid_anchors"] = anchor_rows
    scd = screen[screen["gene"] == "SCD"]
    cms_groups = [scd.loc[scd["CMS_prediction"] == cms, "LFC"].dropna() for cms in ("CMS1", "CMS2", "CMS3", "CMS4")]
    results["organoid_scd_cms"] = {
        "counts": [int(len(x)) for x in cms_groups],
        "medians": [float(x.median()) for x in cms_groups],
        "kruskal_p": float(stats.kruskal(*cms_groups).pvalue),
    }
    msi_tests = []
    for gene in sorted(panel["gene"]):
        subset = screen[screen["gene"] == gene]
        msi = subset.loc[subset["msStatus"] == "MSI", "LFC"].dropna()
        mss = subset.loc[subset["msStatus"] == "MSS", "LFC"].dropna()
        test = stats.mannwhitneyu(msi, mss, alternative="two-sided", method="asymptotic")
        msi_tests.append({"gene": gene, "n_msi": len(msi), "n_mss": len(mss), "p": float(test.pvalue)})
    q = bh_adjust([x["p"] for x in msi_tests])
    for row, value in zip(msi_tests, q):
        row["q"] = float(value)
    results["organoid_msi"] = {
        "n_msi": int(ann["msStatus"].eq("MSI").sum()),
        "n_mss": int(ann["msStatus"].eq("MSS").sum()),
        "genes": {x["gene"]: x for x in msi_tests if x["gene"] in {"SCD", "EGFR"}},
        "n_fdr_lt_005": int(sum(x["q"] < 0.05 for x in msi_tests)),
    }

    dep = root / "analysis" / "depmap_26Q1"
    cross = pd.read_csv(dep / "depmap_organoid_cross_platform_panel.csv")
    full = stats.spearmanr(cross["median_organoid"], cross["median_crc_2d"])
    legacy = cross[cross["gene"] != "CD274"]
    legacy_test = stats.spearmanr(legacy["median_organoid"], legacy["median_crc_2d"])
    context = pd.read_csv(dep / "depmap_target_context_summary.csv").set_index("gene")
    results["cross_platform"] = {
        "n_full": int(len(cross)), "rho_full": float(full.statistic), "p_full": float(full.pvalue),
        "n_legacy": int(len(legacy)), "rho_legacy": float(legacy_test.statistic), "p_legacy": float(legacy_test.pvalue),
        "anchors": {
            gene: {k: float(context.loc[gene, k]) for k in (
                "median_crc_2d", "pct_crc_lt_minus_1", "median_difference_crc_minus_other",
                "median_difference_ci95_lo", "median_difference_ci95_hi", "mannwhitney_p_crc_vs_other"
            )} for gene in ("SCD", "EGFR")
        },
    }

    codep = root / "analysis" / "scd_codependency_qc"
    rank = pd.read_csv(codep / "scd_codependency_qc_adjusted.csv")
    gsea = pd.read_csv(codep / "scd_hallmark_raw_vs_qc.csv")
    targets = rank[rank["gene"].isin(["VPS72", "CDX1", "WLS"])].set_index("gene")
    results["scd_codependency"] = {
        "n_genes": int(len(rank)),
        "n_gene_fdr_lt_005": int((rank["FDR_BH"] < 0.05).sum()),
        "target_rows": targets.to_dict(orient="index"),
        "n_hallmark_fdr_lt_005": int((gsea["FDR_BH"] < 0.05).sum()),
        "dna_repair": gsea[gsea["pathway"] == "HALLMARK_DNA_REPAIR"].set_index("ranking").to_dict(orient="index"),
        "cholesterol_homeostasis": gsea[gsea["pathway"] == "HALLMARK_CHOLESTEROL_HOMEOSTASIS"].set_index("ranking").to_dict(orient="index"),
    }

    gse = pd.read_csv(root / "analysis" / "gse39582_recalculation" / "gse39582_module_clinical.tsv", sep="\t")
    gse_univ = cox_fit(gse, "rfs_months", "rfs_event", ["module_z"])
    gse_adj = cox_fit(gse, "rfs_months", "rfs_event", ["module_z", "age", "mmrd", "chemotherapy_yes", "stage_2", "stage_3", "stage_4"])
    results["gse39582"] = {"rfs_univariable": gse_univ, "rfs_adjusted": gse_adj, "os_analysable": False}

    tcga = pd.read_csv(root / "analysis" / "tcga_cbioportal_592" / "tcga_cbioportal_module_clinical.tsv", sep="\t")
    tcga_univ = cox_fit(tcga, "os_months", "os_event", ["module_z"])
    tcga_stage = pd.get_dummies(tcga, columns=["stage_num"], prefix="stage", dtype=float)
    stage_columns = [column for column in ("stage_2.0", "stage_3.0", "stage_4.0") if column in tcga_stage]
    tcga_adj = cox_fit(tcga_stage, "os_months", "os_event", ["module_z", "age", *stage_columns])
    results["tcga"] = {
        "expression_n": int(tcga["module_z"].notna().sum()),
        "os_univariable": tcga_univ,
        "os_age_stage_adjusted": tcga_adj,
    }

    leuko = root / "analysis" / "gse146771"
    rates = pd.read_csv(leuko / "gse146771_candidate_rates_by_patient_tissue_lineage.csv")
    all_rates = rates[rates["lineage"] == "All leukocytes"]
    paired = all_rates.pivot(index="Sample", columns="Tissue", values="rate").dropna(subset=["N", "T"])
    differences = (paired["T"] - paired["N"]).to_numpy()
    tissue = pd.read_csv(leuko / "gse146771_candidate_tissue_descriptive.csv").set_index("tissue")
    sensitivity = pd.read_csv(leuko / "gse146771_threshold_sensitivity.csv")
    results["gse146771"] = {
        "paired_patients": int(len(paired)),
        "mean_difference_T_minus_N": float(differences.mean()),
        "exact_sign_flip_p": exact_sign_flip_p(differences),
        "tissue_counts": tissue[["n_cells", "candidates", "rate"]].to_dict(orient="index"),
        "threshold_candidate_range": {
            str(threshold): [int(group["total_candidates"].min()), int(group["total_candidates"].max())]
            for threshold, group in sensitivity.groupby("CEACAM5_TPM_threshold")
        },
    }

    g178 = pd.read_csv(root / "analysis" / "gse178341_recalculation" / "gse178341_patient_scores.tsv", sep="\t")
    cms4 = g178["prediction_raw"] == "CMS4"
    cms_whole = stats.mannwhitneyu(g178.loc[cms4, "module_all"], g178.loc[~cms4, "module_all"], alternative="two-sided")
    cms_epi = stats.mannwhitneyu(g178.loc[cms4, "module_Epi"], g178.loc[~cms4, "module_Epi"], alternative="two-sided")
    correlations = {}
    for a, b in (("module_all", "TGFb_all"), ("module_all", "EMT_all"), ("module_Epi", "TGFb_Epi"), ("module_Epi", "EMT_Epi"), ("module_all", "T_cell_fraction_all"), ("module_Epi", "T_cell_fraction_all")):
        test = stats.spearmanr(g178[a], g178[b], nan_policy="omit")
        correlations[f"{a}_vs_{b}"] = {"rho": float(test.statistic), "p": float(test.pvalue)}
    mmrd = g178["MMRStatus"] == "MMRd"
    mmr_tests = {}
    for column in ("T_cell_fraction_all", "exhausted_label_fraction_T", "exhaustion_expression_TNKILC"):
        test = stats.mannwhitneyu(g178.loc[mmrd, column], g178.loc[~mmrd, column], alternative="two-sided")
        mmr_tests[column] = {
            "median_mmrd": float(g178.loc[mmrd, column].median()),
            "median_mmrp": float(g178.loc[~mmrd, column].median()),
            "p": float(test.pvalue),
        }
    q = bh_adjust([x["p"] for x in mmr_tests.values()])
    for row, value in zip(mmr_tests.values(), q):
        row["q"] = float(value)
    results["gse178341"] = {
        "patients": int(g178["PID"].nunique()),
        "mmrd": int(mmrd.sum()), "mmrp": int((~mmrd).sum()),
        "cms_raw_counts": g178["prediction_raw"].value_counts().to_dict(),
        "cms4_whole_p": float(cms_whole.pvalue), "cms4_epithelial_p": float(cms_epi.pvalue),
        "correlations": correlations, "mmr_tests": mmr_tests,
    }

    output = args.outdir / "v7_key_results_recalculated.json"
    output.write_text(json.dumps(results, indent=2, ensure_ascii=False), encoding="utf-8")
    print(json.dumps(results, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
